src/fcmp_pp/fcmp_pp_rust/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+    err: *const E,
+}
+impl<T, E> CResult<T, E> {
+    fn ok(value: T) -> Self {
+        CResult {
+            value: Box::into_raw(Box::new(value)),
+            err: core::ptr::null(),
+        }
+    }
+    fn err(err: E) -> Self {
+        CResult {
+            value: core::ptr::null(),
+            err: Box::into_raw(Box::new(err)),
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn hash_grow_helios(
+    existing_hash: HeliosPoint,
+    offset: usize,
+    existing_child_at_offset: HeliosScalar,
+    new_children: HeliosScalarSlice,
+) -> CResult<HeliosPoint, ()> {
+    let hash = hash_grow(
+        HELIOS_GENERATORS(),
+        existing_hash,
+        offset,
+        existing_child_at_offset,
+        new_children.into(),
+    );
+
+    if let Some(hash) = hash {
+        CResult::ok(hash)
+    } else {
+        // TODO: return defined error here: https://github.com/monero-project/monero/pull/9436#discussion_r1720477391
+        CResult::err(())
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn hash_trim_helios(
+    existing_hash: HeliosPoint,
+    offset: usize,
+    children: HeliosScalarSlice,
+    child_to_grow_back: HeliosScalar,
+) -> CResult<HeliosPoint, ()> {
+    let hash = hash_trim(
+        HELIOS_GENERATORS(),
+        existing_hash,
+        offset,
+        children.into(),
+        child_to_grow_back,
+    );
+
+    if let Some(hash) = hash {
+        CResult::ok(hash)
+    } else {
+        // TODO: return defined error here: https://github.com/monero-project/monero/pull/9436#discussion_r1720477391
+        CResult::err(())
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn hash_grow_selene(
+    existing_hash: SelenePoint,
+    offset: usize,
+    existing_child_at_offset: SeleneScalar,
+    new_children: SeleneScalarSlice,
+) -> CResult<SelenePoint, ()> {
+    let hash = hash_grow(
+        SELENE_GENERATORS(),
+        existing_hash,
+        offset,
+        existing_child_at_offset,
+        new_children.into(),
+    );
+
+    if let Some(hash) = hash {
+        CResult::ok(hash)
+    } else {
+        // TODO: return defined error here: https://github.com/monero-project/monero/pull/9436#discussion_r1720477391
+        CResult::err(())
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn hash_trim_selene(
+    existing_hash: SelenePoint,
+    offset: usize,
+    children: SeleneScalarSlice,
+    child_to_grow_back: SeleneScalar,
+) -> CResult<SelenePoint, ()> {
+    let hash = hash_trim(
+        SELENE_GENERATORS(),
+        existing_hash,
+        offset,
+        children.into(),
+        child_to_grow_back,
+    );
+
+    if let Some(hash) = hash {
+        CResult::ok(hash)
+    } else {
+        // TODO: return defined error here: https://github.com/monero-project/monero/pull/9436#discussion_r1720477391
+        CResult::err(())
+    }
+}
+
+// https://github.com/rust-lang/rust/issues/79609
+#[cfg(all(target_os = "windows", target_arch = "x86"))]
+#[no_mangle]
+pub extern "C" fn _Unwind_Resume() {}
```

Comment:
If mingw-w64 was compiled with dwarf2 exceptions enabled this will cause a multiple definitions error. I [do this](https://github.com/monero-project/monero/pull/9440/commits/866674d5c352093b9a17c1a979575229e74ac631) in #9440, because if you cross-compile the rust standard library with sjlj mingw-w64, you run into the same undefined references to unwind issue. Building std without unwinding should be possible, but I ran into problems. Removing these lines here breaks CI, because distros package sjlj mingw-w64.

I'll try to come up with the least hacky way to work around this. IMO, we could consider dropping the 32-bit Windows target. Windows 11 requires 64-bit and I can't imagine anyone using a >18 year old CPU to run Windows 10.